Allow to use quelpa-use-package as source of truth for quelpa-cache

kiennq opened the pull request:
This allows the `quelpa-cache` to be populated from the usage of `quelpa-use-package` only.
Since we're using `quelpa-cache` as a way to indicate which packages are managed by `quelpa`, this change allows those packages are correctly populated via the config files only.

Also using a simpler form instead of `package-installed-p` to check for package availability so that we don't need to load the whole `package` package when running from init file, improve Emacs init time

kiennq committed:
Allow to use quelpa-use-package as source of truth for quelpa-cache

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
+(defvar quelpa-cache)
+(defvar quelpa-persistent-cache-p)
+
+(when quelpa-use-package-as-source-of-truth
+  (setq quelpa-persistent-cache-p nil))
```
This form belongs in the `defcustom`'s setter so it will take effect when the option is changed.

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
+  :group 'quelpa-use-package)
+
+(defcustom quelpa-use-package-as-source-of-truth nil
+  "If non-nil, only packages installed via this will be managed by `quelpa'.
```
It may be unclear what "this" refers to.

As well, it seems unclear what "managed by Quelpa" means.

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
+(defgroup quelpa-use-package nil
+  "Quelpa handler for `use-package'."
+  :prefix "quelpa-use-package"
+  :group 'use-package)
```
Should the parent group be `use-package` or `quelpa`?

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
-(defvar quelpa-use-package-inhibit-loading-quelpa nil
+(defgroup quelpa-use-package nil
+  "Quelpa handler for `use-package'."
+  :prefix "quelpa-use-package"
```
It's not harmful to specify the prefix, but it's unnecessary for options specified later in the same file.

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
+  :group 'use-package)
+
+(defcustom quelpa-use-package-inhibit-loading-quelpa nil
   "If non-nil, `quelpa-use-package' will do its best to avoid
```
The first line of a docstring should be a complete sentence.  If this one's is being updated, it might as well be fixed.

alphapapa commented on quelpa-use-package.el:
```diff
@@ -37,17 +37,35 @@
+  :type 'boolean
+  :group 'quelpa-use-package)
+
+(defcustom quelpa-use-package-as-source-of-truth nil
```
The name of this variable seems unclear, and a bit mysterious.  Could it be given a more descriptive one?

alphapapa reviewed (changes requested):
I've left some comments on individual lines.  Beyond those, please write a bit about the purpose of this PR and why it's necessary, to help me understand it.

kiennq commented:
> I've left some comments on individual lines. Beyond those, please write a bit about the purpose of this PR and why it's necessary, to help me understand it.

Thanks, let me clarify the purpose of this PR before addressing your comments.
Currently, `quelpa` is marking which packages it's managed via `quelpa-cache`, which is also used to store the custom recipes as well. The cache will be updated when we install a new package via quelpa, or get restored from the cache file if we specify the `quelpa-persistent-cache-p`.

So, if we don't install any new package with `quelpa` in this Emacs session and don't have `quelpa-persistent-cache-p` on, running `quelpa-upgrade-all` will do nothing (besides upgrading `quelpa`).
On other hand, if we have unnecessary packages in the `quelpa-cache`, doing upgrades will cost a lot of time since we're doing repo upgrades for all of those packages as well.

Since we have this `:quelpa` syntax with `use-package`, I think we can consider all packages that are declared with `:quelpa` in the config file as implicitly an entry in the `quelpa-cache`.
Using that and turn off `quelpa-persistent-cache-p` will make us have a clean `quelpa-cache` every time Emacs starts, with the exact information as specified in the config file.

kiennq committed:
check quelpa package availability using a simpler form that doesn't require package
